scarb/src/core/registry/client/http.rs

```diff
@@ -146,12 +151,77 @@ impl<'c> RegistryClient for HttpRegistryClient<'c> {
     }
 
     async fn supports_publish(&self) -> Result<bool> {
-        // TODO(mkaput): Publishing to HTTP registries is not implemented yet.
-        Ok(false)
+        let index_config = self.index_config.load().await?;
+        match index_config.upload {
+            Some(_) => Ok(true),
+            None => Ok(false),
+        }
     }
 
-    async fn publish(&self, _package: Package, _tarball: FileLockGuard) -> Result<()> {
-        todo!("Publishing to HTTP registries is not implemented yet.")
+    async fn publish(&self, package: Package, tarball: FileLockGuard) -> Result<RegistryUpload> {
+        let auth_token =
+            env::var("SCARB_AUTH_TOKEN").map_err(|_| anyhow!("Missing authentication token."))?;
+
+        let path = tarball.path().to_owned();
+        // we need to drop, because windows file locking is very strict
+        drop(tarball);
```

Comment:
You can convert `FileLockGuard` into `AsyncFileLockGuard` which you can deref into `tokio::fs::File` which is what you eventually want I think?

`LocalRegistry` uses `fsx::copy` which makes a hard link on fs instead of copying data -- saves user storage. Here we have a different use-case.

Reply:
Don't thin deref will work, as stream needs an object BUT this is actually very cool idea! I have added an `into_file` to async lock to assume the lock ad used it alongside the idea of converting it to async and it seems to be working :D thanks for this pointer, really appreciate this 🙏 been trying to figure this out longer than I should 😅 

Added change in: https://github.com/software-mansion/scarb/pull/1570/commits/6c399981822e0417f0bcf18ce91b3e5e134efe80